src/shared/helpers.ts

```diff
@@ -88,6 +88,48 @@ export function debounceAsync<T extends unknown[], R extends Promise<unknown>>(
   }
 }
 
+/**
+ * Debounce a function, while allowing the queued arguments to be reduced before
+ * the function is called. With args reducer, we can call the debounced function
+ * with first/last/merged arguments etc.
+ *
+ * @example
+ * ```ts
+ * const debounceWithQueue = new DebounceWithQueue(
+ *   (total: number) => saveToStorage(total),
+ *   (collectedArgs) => collectedArgs.reduce(total, [val] => total + val, 0),
+ *   wait
+ * )
+ * debounceWithQueue.enqueue(10)
+ * debounceWithQueue.enqueue(15)
+ * // results in saveToStorage(25)
+ * ```
+ */
+export class DebounceWithQueue<Args extends unknown[]> {
```

Comment:
What happens if the user closes the browser and we still have items in the queue?

Reply:
They may get lost (upto a second of queue), but we can get latest data on next payment.

Anyway, I was testing this out and will move it into a separate PR when we store spent/balances on payment.